### PR TITLE
Calculate maximum CP

### LINKED ISSUE #9: Add Max CP information
Base Stats are already included in the pokemon data, so Max CP is a relatively simple calculation away. Slight change to the class PokemonMapper based on [the fine work of The Silph Road redditors](https://www.reddit.com/r/TheSilphRoad/comments/4t7r4d/exact_pokemon_cp_formula/), and you should be done.

### PR DESCRIPTION
**Pull request by BrunnerLivio**
I added a max CP Calculator. Source for the calculation from [/u/CpMultiplier](https://www.reddit.com/r/TheSilphRoad/comments/4t7r4d/exact_pokemon_cp_formula/)
Fixes #9

### CODE REVIEW

**Comment by BrunnerLivio** [comment]
@proegssilb The maxCP values are not correct at the moment. I think the problem is I did not add "IndividualAttack", "IndividualStamina" or "IndividualDefense" as described [on reddit](https://www.reddit.com/r/TheSilphRoad/comments/4t7r4d/exact_pokemon_cp_formula/). I don't know how I get these values. Do you know that? 

Also, can you go through cp-calculator.ts, if I calculated it correctly?

File: src/core/cp-calculator/cp-calculator.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        let attack = baseAttack * AppSettings.CP_MULTIPLIER;
+        // Defense = (BaseDefense + IndividualDefense) * TotalCpMultiplier
+        let defense = baseDefense = AppSettings.CP_MULTIPLIER;
+
+        // CP = MAX(10, FLOOR(Stamina^0.5 * Attack * Def^0.5 / 10))
+        return Math.max(
+            Math.floor(
+                Math.pow(stamina, 0.5) * attack * Math.pow(defense, 0.5) / 10
+            )
+        )
```
Comment (proegssilb) [inline src/core/cp-calculator/cp-calculator.ts]: Are you calling Math.max with one argument?

**Review by proegssilb — COMMENTED** [review]
All of the individual values are 0-15. So, for Max CP, you just set it to 15. 

Aside from calling MAX with one argument, and the Individual Stam/Att/Def (AKA IVs), looks great.

**Comment by proegssilb** [comment]
Code speaks louder than words: https://github.com/proegssilb/pokemon-go-rankings/blob/master/generate-output.py#L48

I didn't implement the "max(..., 10)" bit there because I didn't care too much about it. The CP multiplier is brute-forced, because I didn't know about the reddit community's CPM table at the time.

**Comment by BrunnerLivio** [comment]
@proegssilb still not getting the expected results. (Source for maxcp: http://www.imore.com/pokemon-go-gen-2)

Am I still missing something?

File: src/components/pokemon/pokemon.mapper.ts [hunk 2]
```diff
@@ -88,6 +89,9 @@ export class PokemonMapper {
             shoulderModeScale: pkmStgs.camera.shoulderModeScale
         };
 
+        pokemon.maxCP = CPCalculator
+            .Calculate(pokemon.stats.baseStamina, pokemon.stats.baseAttack, pokemon.stats.baseStamina);
```
Comment (proegssilb) [inline src/components/pokemon/pokemon.mapper.ts]: Shouldn't this be using `pokemon.stats.baseDefense` somewhere?

File: src/core/cp-calculator/cp-calculator.spec.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { expect } from 'chai';
+import { CPCalculator } from './cp-calculator';
+
+describe('Class: CPCalculator', () => {
+    it('should output the correct values', () => {
+        // Source: http://www.imore.com/pokemon-go-gen-2
+        let cp: number;
+        // Bayleef
+        cp = CPCalculator.Calculate(122, 155, 120);
```
Comment (proegssilb) [inline src/core/cp-calculator/cp-calculator.spec.ts]: This looks like the wrong argument order to me. I got the right answer with `cp = CPCalculator.Calculate(120, 122, 155);`.

File: src/core/cp-calculator/cp-calculator.spec.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+        expect(cp).to.equal(1296);
+
+    });
+});
```
Comment (proegssilb) [inline src/core/cp-calculator/cp-calculator.spec.ts]: Can we add a test case for another pokemon or two? Here's one for Meganium:

it('should output the correct values for Meganium', () => {
        // source: https://rankedboost.com/pokemon-go/generation-2-tier-list/
        let cp: number;
        cp = CPCalculator.Calculate(160, 168, 202);
        expect(cp).to.equal(2227);
    });

**Review by proegssilb — CHANGES_REQUESTED** [review]
Looks like just a couple of issues with how you call the CP Calculator. See my code comments.

EDIT: Maybe you could pass in a Pokemon object or a PokemonStats object instead of the raw stats? I'm not sure what the right amount of coupling is here.

File: src/components/pokemon/pokemon.mapper.ts (hunk 2 above)
Comment (BrunnerLivio) [reply src/components/pokemon/pokemon.mapper.ts]: Oh, stupid mistake.

**Comment by BrunnerLivio** [comment]
@proegssilb Not sure if we should use `PokemonStats` interface as parameter, because that would mean I need to import inside `core/cp-calculator` from `components/pokemon`. I don't think "core" should depend on "components".

We could fix this by rearranging the structure. But I'm not quite sure if is should move cp-calculator to the components folder.

**Comment by proegssilb** [comment]
I'm not hugely interested in rearranging structure. Passing in PokemonStats is just a nicety.

**Review by proegssilb — APPROVED** [review]

**Pull request merged**